### PR TITLE
from_array: special handling for ndarray, list, and scalar types

### LINKED ISSUE #3556: Simplify graph produced by da.from_array
```
>>> dict(da.from_array(np.array([1, 2]), chunks=2).__dask_graph__())
{
    ('array-de932becc43e72c010bc91ffefe42af1', 0): (<function dask.array.core.getter(a, b, asarray=True, lock=None)>, 'array-original-de932becc43e72c010bc91ffefe42af1', (slice(0, 2, None),)),
    'array-original-de932becc43e72c010bc91ffefe42af1': array([1, 2])
}

>>> dict(da.from_array(np.array([[1, 2], [3, 4]]), chunks=(1, 2)).__dask_graph__())
{
    ('array-7cb83fc33e4505ed4b89aa67e13272f7', 0, 0): (<function dask.array.core.getter(a, b, asarray=True, lock=None)>, 'array-original-7cb83fc33e4505ed4b89aa67e13272f7', (slice(0, 1, None), slice(0, 2, None))),
    ('array-7cb83fc33e4505ed4b89aa67e13272f7', 1, 0): (<function dask.array.core.getter(a, b, asarray=True, lock=None)>, 'array-original-7cb83fc33e4505ed4b89aa67e13272f7', (slice(1, 2, None), slice(0, 2, None))),
    'array-original-7cb83fc33e4505ed4b89aa67e13272f7': array([[1, 2], [3, 4]])
}
```
The above two should really just produce:
```
>>> dict(da.from_array(np.array([1, 2]), chunks=2).__dask_graph__())
{('array-de932becc43e72c010bc91ffefe42af1', 0): array([1, 2])}

>>> dict(da.from_array(np.array([[1, 2], [3, 4]]), chunks=(1, 2)).__dask_graph__())
{
    ('array-7cb83fc33e4505ed4b89aa67e13272f7', 0, 0): (<function _operator.getitem>, 'array-original-7cb83fc33e4505ed4b89aa67e13272f7', (slice(0, 1, None))),
    ('array-7cb83fc33e4505ed4b89aa67e13272f7', 1, 0): (<function _operator.getitem>, 'array-original-7cb83fc33e4505ed4b89aa67e13272f7', (slice(1, 2, None))),
    'array-original-7cb83fc33e4505ed4b89aa67e13272f7': array([[1, 2], [3, 4]])
}
```

### PR DESCRIPTION
**Pull request by crusaderky**
Closes #3556

List, tuple, and all scalar types are now automatically converted to ndarray - just like it happens in all numpy functions.
Please let me know if anybody cares about any weird use cases involving memoryview or list/tuple subclasses; right now I'm just converting them to ndarray.

ndarray with a single chunk produces a minimal 1-key dask dict.
ndarray with more than one chunk uses operator.getitem instead of getter.

### CODE REVIEW

**Comment by crusaderky** [comment]
@mrocklin ready for review

File: dask/array/core.py [hunk 1]
```diff
@@ -2176,6 +2176,9 @@ def from_array(x, chunks, name=None, lock=False, asarray=True, fancy=True,
 
     >>> a = da.from_array(x, chunks=(1000, 1000), lock=True)  # doctest: +SKIP
     """
+    if isinstance(x, (list, tuple, memoryview) + np.ScalarType):
```
Comment (shoyer) [inline dask/array/core.py]: I had no idea `np.ScalarType` was a thing. It looks handy!

**Comment by crusaderky** [comment]
this random instability of travis is killing me... 3 attempts in a row before a good build!

**Comment by mrocklin** [comment]
> this random instability of travis is killing me... 3 attempts in a row before a good build!

I agree that random failures on CI is frustrating.  I recommend opening up a separate issue or, better yet, a PR if you have the time.

**Comment by mrocklin** [comment]
@shoyer are you ok with these changes?  You say that you're not familiar with the array optimizations used here, but at the same time you're probably the person who is the most familiar, and also the person who is the most affected.  OK to merge?

File: dask/array/tests/test_array_core.py [hunk 2]
```diff
@@ -1991,39 +1991,91 @@ def __init__(self, x):
         self.x = x
         self.dtype = x.dtype
         self.shape = x.shape
+        self.ndim = len(x.shape)
```
Comment (shoyer) [inline dask/array/tests/test_array_core.py]: is this actually necessary? you only access `ndim` on numpy arrays

**Comment by shoyer** [comment]
@mrocklin I think you and @jcrist were the main authors of slice optimizations in dask.array. If either of you has time to take a quick look at this, it would be appreciated.

**Comment by mrocklin** [comment]
I haven't looked at them recently enough to have much understanding in my head.  I depend on the tests to ensure that none of the cases we care about for projects like XArray get broken.  

I believe that @jcrist is mostly off doing Yarn things at the moment and may not review things like this.

**Comment by mrocklin** [comment]
I'm also doing what I can to keep up with review, but our incoming PRs / reviewers ratio seems to have increased substantially over the last month or so.

**Comment by shoyer** [comment]
I agree that the test suite seems to cover `getitem` as least as carefully as `getter`, so +1 from me:
https://github.com/dask/dask/blob/master/dask/array/tests/test_optimization.py

**Comment by mrocklin** [comment]
+1 from @shoyer , merging.  Thanks @crusaderky !  Sorry for the long delay.

**Pull request merged**